docs/designs/concepts.md

```diff
@@ -0,0 +1,116 @@
+# Concepts Design Doc
+
+This document describes core concepts used in ElasticDL, and discusses their representations.
+
+## Core Concepts
+
+Fundamentally, machine learning is using algorithms to extract information from raw data and represent it in some type of model.
+
+There two kinds of parameters in a model, one is dense parameter, the other is sparse parameter. The sparse parameter is used by the embedding layer, so it is also called embedding table.
+
+Both dense parameter and embedding table includes a tensor data field, and some other auxiliary fields. The auxiliary fields indluce a name. Embedding table has a extra auxiliary field, indices field.
+
+Tensor is used to represented a n-dim data. There are many data types, such as int8/float32.
+
+Each kinds of parameter has its own kind of gradient. Dense parameter has dense parameter gradient, embedding table has embedding table gradient.
+
+Let's make a short summary, following is all the core concepts used in ElasticDL:
+
+- Model
+- DenseParam
+- EmbeddingTable
+- Tensor
+- DenseParamGradient
+- EmbeddingTableGradient
+
+## In-Memory Representation
+
+
+## Message Representation
+
+```proto
+enum TensorDtype {
+  // Not a legal value for DataType. Used to indicate a DataType field
+  // has not been set.
+  DT_INVALID = 0;
+
+  DT_INT8 = 1;
+  DT_INT16 = 2;
+  DT_INT32 = 3;
+  DT_INT64 = 4;
+  DT_FLOAT16 = 5;
+  DT_FLOAT32 = 6;
+  DT_FLOAT64 = 7;
+  DT_BOOL = 8;
+}
+
+message Tensor {
+  repeated int64 dims = 1;
+  bytes content = 2;
+  TensorDtype dtype = 3;
+}
+
+message DenseParam {
+  string name = 1;
+  Tensor tensor = 2;
+}
+
+message EmbeddingTable {
+  string name = 1;
+  Tensor tensor = 2;
+  repeated int64 indices = 3;
+}
+
+message Model {
+  bool initialized = 1;
+  int32 version = 2;
```

Comment:
Once step updating increase the version one time. The training of a Bert-large model need 1M steps udpating. So int32 is enough here.